drivers: Add text display interface

thedjnK opened the pull request:
Adds the base driver include file and noritake itron text display dts file for review.

- [X] Base noritake itron driver code
- [X] Base grove LCD driver code
- [X] Base HD44780 driver code
- [x] Complete noritake itron driver code
- [X] Complete grove LCD driver code (unable to test)
- [X] Complete HD44780 driver code
- [X] API
- [x] Documentation
- [x] Mark API is experimental/unstable
- [x] Release notes update
- [ ] (Optional) more advanced sample code e.g. display configuration, add test character
- [ ] (Add as feature issue, not for this PR) remove old grove driver and update samples using it to use new driver
- [ ] (Add as feature issue, not for this PR) remove old HD44780 driver sample

Tested:

- amc2004a (20x4 HD44780-clone): ☑ Works on nucleo_f746zg in both 4 and 8-bit modes
- Noritake Itron gu280x16g (Noritake itron UART): ☑ Works on nucleo_f401re
- seed grove lcd rgb (Jinghua Display JHD1313): ☒ Do not have shield to test with
- Noritake Itron gu24025ecpb (24x2 HD44780-clone): ☑ Works on stm32f4_disco with 800ms boot delay, sometimes displays '6', '66' or '666' at start when board reset button is pressed but this seems to be a quirk of the display/reset or other time value that needs increasing
- Unbranded 1602a (16x2 HD44780-clone): ☑ Works on stm32f4_disco in 4-bit mode (not checking on 8-bit mode as the debugger deciding if it detects a core or not completely at random is almost making me throw it against a wall)

Fixes #51912

thedjnK committed:
drivers: Add auxdisplay (text) interface

Adds the base driver include file and base auxiliary display
dts file.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
drivers: auxdisplay: Add noritake itron VFD auxiliary display

Adds the driver for a Noritake Itron VFD auxiliary display.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+ * @{
+ */
+
+#include <zephyr/device.h>
+#include <stddef.h>
+#include <zephyr/types.h>
```
```
#include <stdint.h>

#include <zephyr/device.h>
```

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+	uint16_t columns;
+	/** Number of character rows */
+	uint16_t rows;
+	/** Display-specific data (e.g. 4-bit or 8-bit mode for HD44780-based displays) */
+	uint8_t mode;
```
nit: make mode an opaque type, similar to what we do for `gpio_flags_t`

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+ *
+ * @retval 0 on success else negative errno code.
+ */
+static inline int text_display_power_set(const struct device *dev, bool enabled)
```
suggestion: naming `text_display_set_enabled(dev, true/false)` IMO it's more clear when reading code.

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+ *
+ * @retval 0 on success else negative errno code.
+ */
+static inline int text_display_cursor_set(const struct device *dev, bool enabled)
```
same: `cursor_set_enabled`

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+ * @brief Set cursor (and write position) on a text display
+ *
+ * @param dev	Pointer to device structure
+ * @param type	Type (todo: exact position or offset from current)
```
not clear what type is

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+ *
+ * @retval 0 on success else negative errno code.
+ */
+static inline int text_display_write(const struct device *dev, const uint8_t *data, uint16_t len)
```
note: since this seems to be an API for apps, needs syscalls

gmarull commented on include/zephyr/drivers/text_display.h:
```diff
@@ -0,0 +1,391 @@
+/**
+ * @brief Text Display Interface
+ * @defgroup text_display_interface Text Display Interface
+ * @ingroup text_display_interfaces
```
this is wrong: `io_interfaces`

gmarull commented on dts/bindings/text_display/noritake,itron.yaml:
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: Noritake Itron VFD
```
binding seems to be in the wrong commit

gmarull commented on dts/bindings/text_display/noritake,itron.yaml:
```diff
@@ -0,0 +1,29 @@
+include: uart-device.yaml
+
+properties:
+    columns:
+        type: int
+        required: true
+
+    rows:
+        type: int
+        required: true
```
if common to all text display APIs, make a common binding file

gmarull commented on dts/bindings/text_display/noritake,itron.yaml:
```diff
@@ -0,0 +1,29 @@
+        type: int
+        required: true
+
+    reset-gpios:
+        type: phandle-array
+
+    busy-gpios:
+        type: phandle-array
+
+    default-font:
+        type: int
```
missing help

gmarull commented on drivers/text_display/text_display_noritake_itron.c:
```diff
@@ -0,0 +1,259 @@
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/text_display.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
```
nit: review includes, this file is not using e.g. pinctrl.

gmarull commented on drivers/text_display/text_display_noritake_itron.c:
```diff
@@ -0,0 +1,259 @@
+#define NORITAKE_ITRON_CHARACTER_CARRIAGE_RETURN 0x0d
+
+
+struct text_display_noritake_itron_data {
+	uint8_t unused;
+};
```
driver data can be `NULL`

gmarull commented on drivers/text_display/text_display_noritake_itron.c:
```diff
@@ -0,0 +1,259 @@
+	int rc;
+
+	if (!device_is_ready(config->uart)) {
+		LOG_ERR("clock control device not ready");
```
copy&paste here

gmarull commented on drivers/text_display/text_display_noritake_itron.c:
```diff
@@ -0,0 +1,259 @@
+	return 0;
+}
+
+static int noritake_itron_position_blinking_set(const struct device *dev, bool enabled)
+{
+	return -ENOTSUP;
+}
+
+static int noritake_itron_cursor_position_set(const struct device *dev, uint8_t type, uint8_t x, uint8_t y)
+{
+	return -ENOTSUP;
+}
+
+static int noritake_itron_display_position_set(const struct device *dev, uint8_t type, uint8_t x, uint8_t y)
+{
+	return -ENOTSUP;
+}
+
+static int noritake_itron_details_get(const struct device *dev, uint16_t *columns, uint16_t *rows, uint8_t *character_set)
+{
+	return -ENOTSUP;
+}
```
if APIs are optional, maybe allow them to be `NULL` and return `-ENOSYS` at interface level.

gmarull reviewed (changes requested):
some initial comments

cfriedt reviewed (changes requested):
I love that you're adding text-based LCDs here!

It's been a while but would be a great addition on my next demo.

Just a couple of questions / comments:
- Are either of grove or itron displays capable of producing interrupts (e.g. via GPIO)? If so, I would suggest building those capabilities into the drivers.
- Some of the drivers sleep within the API calls. In general, it would be better to use a semaphore or spinlock and either delayed work or an ISR to unlock 

Thanks though - I'm really looking forward to seeing this upstream 😁

cfriedt commented:
It might also be useful to wrap API calls in a `__syscall` wrapper (see `<zephyr/drivers/gpio.h>`. For kernel-mode only builds, there is very little overhead (other than some argument checking), but for user mode, it protects access to the driver data

thedjnK committed:
drivers: auxdisplay: Port Jinghua Display JHD1313 driver

Ports the Jinghua Display JHD1313 LCD (with RGB backlight) driver
to use the new auxdisplay driver interface. This driver is used on
the seeed grove LCD RGB display, and replaces it.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented:
> Are either of grove or itron displays capable of producing interrupts (e.g. via GPIO)? If so, I would suggest building those capabilities into the drivers.

@cfriedt Not sure what you mean here by interrupt, unless you mean if they have user input/buttons/touchscreen? The noritake itron has a busy output line which I did expose here but will hide internally in the driver and not expose externally, HD44780 does not have inputs.

cfriedt commented:
> @cfriedt Not sure what you mean here by interrupt, unless you mean if they have user input/buttons/touchscreen? The noritake itron has a busy output line which I did expose here but will hide internally in the driver and not expose externally, HD44780 does not have inputs.

Busy line should do for a gpio interrupt.

I could make some additional code comments if you need clarification, but my general suggestion was to avoid sleeping (ans also busy looping). Sleeping is usually bad for drivers.

Using an interrupt (e.g. a busy line) is one way to avoid sleeping, and another way to avoid sleeping is to use a workqueue. It might be smart to use a dedicated workqueue in this case.

thedjnK commented:
@cfriedt 😆 this was the comment from the architecture meeting:

> text_display_is_busy probably debatable 

I just removed it so I can add it back quickly...

cfriedt commented on drivers/auxdisplay/auxdisplay_grove_lcd.c:
```diff
@@ -0,0 +1,355 @@
+	 * VDD to power on, so pause a little here, 30 ms min, so we go 50
+	 */
+	LOG_DBG("delay 50 ms while the VDD powers on");
+	k_sleep(K_MSEC(50));
```
Probably better to move this to something like an `enable()` function.

thedjnK commented on drivers/text_display/text_display_grove_lcd.c:
```diff
@@ -0,0 +1,355 @@
+
+	rc = i2c_write_dt(&config->bus, clear, sizeof(clear));
+	LOG_DBG("clear, delay 20 ms");
+	k_sleep(K_MSEC(20));
```
This is from the original zephyr driver (copy and paste) and would think sleeping is the only course since the command should be blocking until it is complete (someone raised this in the architecture meeting)

cfriedt replied:
Commands can still be blocking if you use e.g. a semaphore (semaphore can be unlocked via ISR or workqueue).

This is done in a few places in e.g. sensor drivers.

Just a thought.

thedjnK committed:
CODEOWNERS: Add self as aux display maintainer

Add myself as auxiliary display maintainer.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
MAINTAINERS: Add self as aux display maintainer

Add myself as auxiliary display maintainer.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
samples: driver: Add auxdisplay sample

Adds a simple auxdisplay sample which outputs hello world and the
name of the board. A sample overlay is provided for the
nucleo_f746zg board with a Hitachi HD44780-compatible display.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented:
OK I've made some fixes and updates, PM is implemented in my display, added a simple sample application, still some work to do on it.

@cfriedt I'm a bit stuck with using interrupts and semaphores, I was about to switch to them on my display driver, then I stopped and paused because that would pose a problem, a problem for me at least, I want to be able to use the display from a bootloader (i.e. mcuboot) and imposing semaphores then means I would need to build mcuboot with mutlthreading support, just enabling that option without using the functions in code adds 4KB to a bootloader build which isn't a small number, and thinking about it, the delays should not really be long for a display (for mine at 38400 baud, assuming you output 40 characters and every 3 characters you have to pause for 4ms, the total time needed to display all that is about 63ms, and whilst the display is busy, other tasks can run **perhaps**, so at least for my driver (note: noritake display, not the groove) I don't think it's worth changing to interrupts. I don't think it's worth doing either for the groove but that's because that driver is basically a straight up copy and paste from existing zephyr code, I don't have that display to test with so don't want to submit a driver that might be broken and I'm unable to test with. For a HD44780 driver (which will have to come at a later date) I will think about it, although I think that would fall under the same bucket of using `k_sleep()` because I don't think those display drivers have an interrupt line either, it's all "hold these lines for a minimum time period".

cfriedt commented:
> OK I've made some fixes and updates, PM is implemented in my display, added a simple sample application, still some work to do on it.
> 
> @cfriedt I'm a bit stuck with using interrupts and semaphores.

Device Tree can fix that 😄 Make it a DT property and hook up things in your INIT.

nordicjm commented:
@cfriedt not for setting up the GPIO, that's already done in this PR, it's for usage of semaphores, the burden of enabling that is too great for a simple bootloader

cfriedt commented:
> @cfriedt not for setting up the GPIO, that's already done in this PR, it's for usage of semaphores, the burden of enabling that is too great for a simple bootloader

@nordicjm - you can use DT properties to compile-in or compile-out support for semaphores. In the non-semaphore version just use a different function pointer for achieving the same thing. Then you get the best of both worlds :-)

gmarull commented:
> > @cfriedt not for setting up the GPIO, that's already done in this PR, it's for usage of semaphores, the burden of enabling that is too great for a simple bootloader
> 
> @nordicjm - you can use DT properties to compile-in or compile-out support for semaphores. In the non-semaphore version just use a different function pointer for achieving the same thing. Then you get the best of both worlds :-)

Just saw this comment: DT describes hardware, not software configuration ;-)

cfriedt commented:
> > > @cfriedt not for setting up the GPIO, that's already done in this PR, it's for usage of semaphores, the burden of enabling that is too great for a simple bootloader
> > 
> > @nordicjm - you can use DT properties to compile-in or compile-out support for semaphores. In the non-semaphore version just use a different function pointer for achieving the same thing. Then you get the best of both worlds :-)
> 
> Just saw this comment: DT describes hardware, not software configuration ;-)

@gmarull - that's right. In this case the DT would be describing whether or not there are interrupts available for the device (so business as usual) and then the driver would be configured to either use an interrupt-driven routine or polling based on that (also pretty normal, right?).

I'm just pointing out that it can be done in a compile-time const way inside of `DT_INST_FOREACH_STATUS_OKAY()`. It can be done at runtime in init() too of course.

nordicjm commented:
> @gmarull - that's right. In this case the DT would be describing whether or not there are interrupts available for the device (so business as usual) and then the driver would be configured to either use an interrupt-driven routine or polling based on that (also pretty normal, right?).

That doesn't make sense, if a display needs an interrupt then it needs an interrupt and the driver would have it, if it doesn't need an interrupt then it wouldn't be present in the driver. If there is an optional one then it would down to the driver if it used it or not (and with displays, they are not optional, the display I have here will happily print complete garbage if you do not use and respect the busy line). And from a dts perspective, it's just a phandle-array

thedjnK committed:
drivers: auxdisplay: Add Hitachi HD44780 driver

Adds an auxiliary display driver for Hitachi HD44780-based (and
compatible) LCD displays.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented:
Updated the PR, almost in a finished state now, added a HD44780 driver, which is complete and working (not 100% feature complete but that can be improved in a future release), changed the API a bit so need to look at the grove lcd and noritake drivers, then add documentation and mark the API as unstable or experimental. Feel free to take a look/test (ignoring grove and noritake drivers) and let me know your feeback @gmarull @henrikbrixandersen @cfriedt

thedjnK commented on dts/bindings/auxdisplay/hit,hd44780.yaml:
```diff
@@ -0,0 +1,58 @@
+      description: |
+          Array of GPIOs connected to the DB lines of the display, for 8-bit
+          mode the first entry must be for DB0 and final for DB7 (8 entries),
+          for 4-bit mode the first entry must be for DB4 and final for DB7 (4
+          entries)
```
wrong

thedjnK committed:
doc: hardware: peripheral: Add auxdisplay documentation

Adds base auxdisplay documentation.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK committed:
doc: api: Add auxdisplay as unstable API

Adds the new auxiliary display peripheral as an unstable API.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

martinjaeger commented:
It looks like the driver was renamed from `text_display` to `auxdisplay`. What was the reason for that change? A text display can be the main user interface a device has, so I think the previous driver name was more suitable.

nordicjm commented:
> It looks like the driver was renamed from text_display to auxdisplay. What was the reason for that change? A text display can be the main user interface a device has, so I think the previous driver name was more suitable.

To match up with the driver name in linux: https://elixir.bootlin.com/linux/latest/source/drivers/auxdisplay

martinjaeger commented:
> To match up with the driver name in linux: https://elixir.bootlin.com/linux/latest/source/drivers/auxdisplay

Mhh, ok. For a Linux computer this type of display can indeed only be an auxiliary display. I would argue it's more important to have a self-explanatory name for the typical Zephyr use case than matching with the Linux driver name.

henrikbrixandersen commented:
I prefer the auxdisplay name.

cfriedt commented on drivers/auxdisplay/auxdisplay_noritake_itron.c:
```diff
@@ -0,0 +1,378 @@
+#define NORITAKE_ITRON_BUSY_WAIT_LOOPS 125
+
+/* Time values when multithreading is enabled */
+#define NORITAKE_ITRON_BUSY_MAX_TIME K_MSEC(500)
```
This is a very long time to de overhead to wait to learn that a display is dead. I would suggest waiting for a shorter maximum period of time. Not going to block on it though, since (presumably) the mode time to wait for the semaphore when the display is working is much, much smaller.

thedjnK committed:
doc: release: 3.4: Add auxdisplay addition node

Adds a note on unstable auxiliary display support being added,
alongside 3 drivers.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented:
Rebased for 3.4, if anyone has time to review this it would be most appreciated!

cfriedt reviewed (dismissed):
LGTM

kartben reviewed (changes requested):
Your sample (which is perfect, I really like that it demonstrates how to use power management) is missing a README, and a YAML file for Twister.
See https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html for more details, and [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/templates/sample.tmpl) for a README template.

thedjnK commented:
> Your sample (which is perfect, I really like that it demonstrates how to use power management) is missing a README, and a YAML file for Twister.

So the problem with this is that it is not buildable through twister, because there is no in-tree board or shield with an auxdisplay (all my testing has them manually wired up, and it is wired up with individual wires to pins) - suggestions for this @PerMac?

PerMac commented:
> > Your sample (which is perfect, I really like that it demonstrates how to use power management) is missing a README, and a YAML file for Twister.
> 
> So the problem with this is that it is not buildable through twister, because there is no in-tree board or shield with an auxdisplay (all my testing has them manually wired up, and it is wired up with individual wires to pins) - suggestions for this @PerMac?

From this description it sounds like a case for using "fixture": https://docs.zephyrproject.org/latest/develop/test/twister.html#fixtures
There is no list of allowed names. It works just by matching strings from testcase.yaml and hardware map/cli arg. Maybe try greping through testcase/sample.yamls for "fixture:" and see if something matching your case is already there? Twister builds samples with fixtures but then is not executing them if appropriate hw is not found.

PerMac commented:
If the sample is not even buildable I would question how useful it is as a sample. Can this sample be built for a target (e.g. nrf52840dk) if an overlay is added, telling that there is this display connected and defining how? If so I would assume this is the way to go. Provide an example for a particular board with defined pins, add this board to platform_allow and also add fixture telling that such display is needed. Isn't it the way how some samples using gpio_loopback are done? E.g. https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/spi_bitbang

PerMac reviewed (dismissed):
Please make the sample more useful, as described in the above comment.

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+#include <stddef.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/types.h>
```
not needed

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+extern "C" {
+#endif
+
+#define AUXDISPLAY_LIGHT_NOT_SUPPORTED 0
```
not documented?

thedjnK replied:
Documented

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+	AUXDISPLAY_DIRECTION_COUNT,
+};
+
+struct auxdisplay_light {
+	/** Minimum and maximum light levels, will both be 0 if unsupported */
+	uint8_t minimum;
+	uint8_t maximum;
+};
```
this is not documented correctly (struct, each field)

thedjnK replied:
Fixed

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+	auxdisplay_custom_character_set_api custom_character_set;
+	auxdisplay_write_api write;
+	auxdisplay_custom_command_api custom_command;
+};
```
driver stuff is usually hidden under INTERNAL_HIDDEN

cfriedt replied:
I always forget this one myself.

thedjnK replied:
Made hidden

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+ * @param dev		Pointer to device structure
+ * @param enabled	True to enable cursor, false to disable
+ *
+ * @retval		0 on success else negative errno code.
```
`@retval` usage is wrong here, syntax is `@retval code desc`

thedjnK replied:
Fixed up

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+/**
+ * @brief		Set cursor enabled status on a text display
+ *
+ * @param dev		Pointer to device structure
```
Description is ambiguous here, maybe "Aux display device instance"

thedjnK replied:
Updated

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+		return -ENOSYS;
+	}
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	pm_device_state_get(dev, &state);
+
+	if (state == PM_DEVICE_STATE_SUSPENDED) {
+		return -EIO;
+	}
+#endif
```
-1 on this, other options:

1. Devices can implement runtime PM
2. App keeps track of device state (fragile)

thedjnK replied:
Devices do implement PM but I thought it would have been better to put common parts here to check if it was in power down mode because if a display is powered down, you can't change its configuration (and if you can it's not really powered down)

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,824 @@
+	}
+#endif
+
+	return api->position_blinking_set_enabled(dev, enabled);
+}
+
+/**
+ * @brief		Set cursor shift after character write and display shift
+ *
+ * @param dev		Pointer to device structure
+ * @param direction	Sets the direction of the display when characters are written
+ * @param display_shift	If true, will shift the display when characters are written
+ *			(which makes it look like the display is moving, not the cursor)
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int auxdisplay_cursor_shift_set(const struct device *dev,
+					  uint8_t direction, bool display_shift);
+
+static inline int z_impl_auxdisplay_cursor_shift_set(const struct device *dev,
```
I think we should follow the same structure everywhere: `set_enabled` vs. `shift_set`. @carlescufi do we have guidelines on this?

thedjnK replied:
It does follow it really, `auxdisplay_position_blinking_set_enabled_api` = `position blinking` is being `set` if it is `enabled` or not -> `auxdisplay_cursor_shift_set_api` = `cursor shift` is being `set` for the `direction` and `shift` (direction and display_shift would equate to enabled)

gmarull reviewed (changes requested):
did an initial review of the interface

thedjnK commented:
Added overlay to sample, not addressed outstanding comments on code as of yet but will do soon.

MaureenHelm commented:
@gmarull can you take another look please?

nordicjm commented:
@cfriedt @gmarull @kartben can you re-review please?

cfriedt commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,887 @@
+ * @retval	-ENOSYS if not supported/implemented.
+ * @retval	-errno Negative errno code on other failure.
+ */
+__syscall int auxdisplay_is_busy(const struct device *dev);
```
This just seems prone to race conditions, IMHO. Please require synchronization within the driver. If blocking is an issue, add a timeout, and check for isr context.

thedjnK replied:
The only driver that implements it is the noritake and all it does is read the state of a GPIO, I don't see what needs to be sync'd?

cfriedt commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,887 @@
+		return -ENOSYS;
+	}
+
+#ifdef CONFIG_PM_DEVICE
```
I'm not  crazy about all of the pm stuff inline, and would also expect the device to just wake up when I/O happens rather than return an error (but I'm not a pm expert)

thedjnK replied:
Not how PM works with other systems, if the device is suspended it needs to be resumed before any functions will work, I put the PM code here because no driver is able to do anything if the device is in sleep mode rather than duplicating in every auxdisplay driver but maybe this implementation is wrong, comments @gmarull ?

thedjnK replied:
@gmarull can you let me know if this is OK or if it should be changed? This seems like the final thing needed before this can be merged.

cfriedt replied:
I'll mark this as resolved, since I honestly do not deal with PM a lot and there is already input from @gmarull

cfriedt reviewed (changes requested):
Namespacing (filenames, macro and symbol prefixes) could be much shorter.

Typedef'd API functions should probably end in `_t` which i think is more conventional.

Not a big fan of returning an error when a device is asleep. Maybe just wake it up?

thedjnK commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -296,6 +296,7 @@ jdi	Japan Display Inc.
 jedec	JEDEC Solid State Technology Association
 jesurun	Shenzhen Jesurun Electronics Business Dept.
 jianda	Jiandangjing Technology Co., Ltd.
+jinghua-display	Shenzhen Jinghua Displays Electronics Co., Ltd.
```
Fixed

thedjnK commented on dts/bindings/auxdisplay/jinghua-display,jhd1313.yaml:
```diff
@@ -0,0 +1,11 @@
+#
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_noritake_itron.h:
```diff
@@ -0,0 +1,24 @@
+/*
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_hitachi_hd44780.h:
```diff
@@ -0,0 +1,25 @@
+/*
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_hitachi_hd44780.c:
```diff
@@ -0,0 +1,588 @@
+/*
+ * Copyright (c) 2023 Jamie McCrae
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.noritake_itron:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 Jamie McCrae
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.jinghua_display_jhd1313:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 Jamie McCrae
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.hitachi_hd44780:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Jamie McCrae
+# SPDX-License-Identifier: Apache-2.0
```
Fixed

thedjnK commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,887 @@
+ * @typedef	auxdisplay_cursor_set_enabled_api
+ * @brief	Callback API to turn display cursor visibility on or off
+ * See auxdisplay_cursor_set_enabled() for argument description
+ */
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_noritake_itron.c:
```diff
@@ -0,0 +1,378 @@
+#include <zephyr/drivers/uart.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/logging/log.h>
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_noritake_itron.c:
```diff
@@ -0,0 +1,378 @@
+/*
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_jinghua_display_jhd1313.c:
```diff
@@ -0,0 +1,375 @@
+#define REGISTER_G	0x03
+#define REGISTER_B	0x02
+
+struct jinghua_display_jhd1313_data {
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_jinghua_display_jhd1313.c:
```diff
@@ -0,0 +1,375 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(auxdisplay_jinghua_display_jhd1313_lcd_rgb, CONFIG_AUXDISPLAY_LOG_LEVEL);
```
Fixed

thedjnK commented on drivers/auxdisplay/auxdisplay_jinghua_display_jhd1313.c:
```diff
@@ -0,0 +1,375 @@
+/*
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.noritake_itron:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 Jamie McCrae
+# SPDX-License-Identifier: Apache-2.0
+
+config NORITAKE_ITRON
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.jinghua_display_jhd1313:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config JINGHUA_DISPLAY_JHD1313
+	bool "Jinghua Display JHD1313 driver"
```
Fixed

thedjnK commented on drivers/auxdisplay/Kconfig.hitachi_hd44780:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Jamie McCrae
+# SPDX-License-Identifier: Apache-2.0
+
+config HITACHI_HD44780
```
Fixed

gmarull commented on drivers/auxdisplay/auxdisplay_itron.c:
```diff
@@ -0,0 +1,378 @@
+			}
+		}
+	} else {
+#ifdef CONFIG_PM_DEVICE
+		/* Ensure display is powered on */
+		auxdisplay_itron_pm_action(dev, PM_DEVICE_ACTION_RESUME);
+#endif
```
this call must never be invoked directly, but via e.g. runtime PM calls.

nordicjm replied:
The problem is that if the board is powered then puts the display into suspend mode, the only way to wake the screen up is to send the wake up command, which is what this does. So if you press the reset button on the board whilst the display was suspended, without this code it will not be usable because it is still in sleep mode, so the function is called directly without going through the PM system so that it just sends the wake up command. Either that or I need to duplicate the same code in 2 functions.

gmarull replied:
why can't `pm_device_runtime_get(dev)` be called?

nordicjm replied:
The device's PM would show as active, since the board would be starting up? The issue is caused if the device is put into sleep mode and MCU is rebooted, the display keeps its sleep status but zephyr does not know that the display was suspended before it was reset.

gmarull replied:
when using the device runtime framework, init call is expected to leave the device disabled, then, it's only activated as needed at runtime. Ref. https://docs.zephyrproject.org/latest/services/pm/device_runtime.html

nordicjm replied:
I've not used that or seen it used in any drivers, I use `pm_device_action_run`. And that functionality does not really make sense to a display since in sleep mode the display turns off, the application (not the driver) should control the power state of the display, not turning it on whilst functions are called then turning it off again after since there'd be no point in having a display. And this is how other drivers work too, in fact I can only see 10 drivers in the whole zephyr tree using `pm_device_runtime_get`

henrikbrixandersen commented:
Where are we with this PR? It would be nice to have text display support in place for Zephyr v3.4.0.

cfriedt commented:
> @cfriedt I'm a bit stuck with using interrupts and semaphores, I was about to switch to them on my display driver, then I stopped and paused because that would pose a problem, a problem for me at least, I want to be able to use the display from a bootloader (i.e. mcuboot) and imposing semaphores then means I would need to build mcuboot with mutlthreading support, just enabling that option without using the functions in code adds 4KB to a bootloader build which isn't a small 

The issue I have here is that you are "optimizing" the Driver API for your particular use case (operating in single-threaded mode) whereas the vast majority of users of the API would not be using it in single-threaded mode).

The implication, is that this influences the Driver API design in a negative way and would make the API itself quite fragile and error prone for (probably) the vast majority of users. You yourself (I think you've added a MAINTAINTERS entry) would end up fixing bug after bug after bug until the API was reworked anyway.

I'm just suggesting you pay off the tech debt upfront and make life easier for your future self.

You actually can use semaphores in multithreading mode and directly from ISRs as long as the timeout parameter is `K_NO_WAIT`. In that case, return `-EBUSY`, `-EAGAIN`, or `-ETIMEDOUT`.

One aspect of this kind of design is that you could "advertise" it as fully supporting both threaded and non-threaded mode in your API doc.

thedjnK commented:
> Where are we with this PR? It would be nice to have text display support in place for Zephyr v3.4.0.

I had a quick chat with gmarull earlier about power management in zephyr as it's substantially changed since I last used it so need to go through the docs and update the code sometime to work with the new power management style, I'd like to get it in for 3.4 but not sure on available spare time.

> @thedjnK - the issue I have here is that you are "optimizing" the Driver API for your particular use case (operating in single-threaded / bootloader mode) whereas the vast majority of users of the API would likely be using it in threaded mode).

I did change the PR that comment for the noritake to use semaphores if multithreading was enabled if memory serves me right, not sure if I changed the others, I don't want to touch the JHD one because it's not my driver and I don't have that display to test with, it's the one that was a misc driver before, the HD one I will it to add. Though I would say generally an auxdisplay would be used by a single thread whereby other threads pass messages to display or whatnot, it would be the exception for multiple threads to directly use the display.

thedjnK commented:
@cfriedt @gmarull @henrikbrixandersen issues should be addressed, can you review please?

thedjnK commented:
@kartben too

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,819 @@
+#include <stddef.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
```
not needed

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,819 @@
+	/** Number of custom characters supported by display (0 if unsupported) */
+	uint8_t custom_characters;
+
+	/** Width and height (in pixels) of a custom character, supplied custom characters should
+	 *  match these requirements, both 0 when unsupported.
+	 */
+	uint8_t custom_character_width;
+	uint8_t custom_character_height;
```
this only documents 1 field

gmarull commented on include/zephyr/drivers/auxdisplay.h:
```diff
@@ -0,0 +1,819 @@
+	uint8_t character_code;
+};
+
+/**
+ * @typedef	auxdisplay_display_on_t
+ * @brief	Callback API to turn display on
+ * See auxdisplay_display_on() for argument description
+ */
+typedef int (*auxdisplay_display_on_t)(const struct device *dev);
```
this should also be under internal hidden

gmarull commented on drivers/auxdisplay/auxdisplay_itron.h:
```diff
@@ -0,0 +1,24 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef H_AUXDISPLAY_ITRON_
+#define H_AUXDISPLAY_ITRON_
+
+#define AUXDISPLAY_ITRON_BRIGHTNESS_MIN 1
+#define AUXDISPLAY_ITRON_BRIGHTNESS_MAX 8
+
+#define AUXDISPLAY_ITRON_CHARACTER_MIN 0x20
+#define AUXDISPLAY_ITRON_CHARACTER_BACK_SPACE 0x08
+#define AUXDISPLAY_ITRON_CHARACTER_TAB 0x09
+#define AUXDISPLAY_ITRON_CHARACTER_LINE_FEED 0x0a
+#define AUXDISPLAY_ITRON_CHARACTER_CARRIAGE_RETURN 0x0d
+
+enum {
+	AUXDISPLAY_ITRON_MODE_UNKNOWN = 0,
+	AUXDISPLAY_ITRON_MODE_UART,
+};
```
unless this is shared with other modules, this header can be removed

thedjnK replied:
it can be used with the capabilities function

gmarull commented on drivers/auxdisplay/auxdisplay_itron.c:
```diff
@@ -0,0 +1,451 @@
+		.busy_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, busy_gpios, {0}),			\
+		.reset_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, {0}),			\
+	};											\
+	PM_DEVICE_DT_INST_DEFINE(inst, auxdisplay_itron_pm_action);				\
```
leftover?

gmarull commented on drivers/auxdisplay/Kconfig.itron:
```diff
@@ -0,0 +1,9 @@
+config AUXDISPLAY_ITRON
+	bool "Noritake Itron VFD driver"
+	default y
+	depends on DT_HAS_NORITAKE_ITRON_ENABLED
```
`select GPIO`, `select SERIAL`

gmarull commented on drivers/auxdisplay/auxdisplay_itron.c:
```diff
@@ -0,0 +1,451 @@
+#include <zephyr/drivers/auxdisplay.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
```
leftover?

gmarull commented on drivers/auxdisplay/Kconfig.jhd1313:
```diff
@@ -0,0 +1,9 @@
+config AUXDISPLAY_JHD1313
+	bool "Jinghua Display JHD1313 driver"
+	default y
+	depends on DT_HAS_JHD_JHD1313_ENABLED
```
`select I2C`

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/auxdisplay.h>
+#include <zephyr/pm/device.h>
```
leftover

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+	struct i2c_dt_spec bus;
+};
+
+static uint8_t colour_define[][4] = {
```
`const`

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+	{ 0,   0,   255 },      /* Blue */
+};
+
+static void auxdisplay_jhd1313_reg_set(const struct device *i2c, uint8_t addr, uint8_t data)
+{
+	uint8_t command[2] = { addr, data };
+
+	i2c_write(i2c, command, sizeof(command), JHD1313_BACKLIGHT_ADDR);
+}
```
any reason to not use i2c_dt_spec?

thedjnK replied:
as above

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+	}
+
+	if (y == 0U) {
+		x |= 0x80;
+	} else {
+		x |= 0xC0;
```
magic constants?

thedjnK replied:
as above

thedjnK replied:
Moved to defines

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+
+	rc = i2c_write_dt(&config->bus, clear, sizeof(clear));
+	LOG_DBG("Clear, delay 20 ms");
+	k_sleep(K_MSEC(20));
```
magic sleep?

thedjnK replied:
as above

thedjnK replied:
Moved to defines

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+	rc = i2c_write_dt(&config->bus, buf, sizeof(buf));
+
+	LOG_DBG("Set display_state options, delay 5 ms");
+	k_sleep(K_MSEC(5));
```
magic sleep

thedjnK replied:
Not my driver, I just took what was in tree already and have no way to test it

thedjnK replied:
Moved to defines

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+		.cursor = false,							\
+		.blinking = false,							\
+	};										\
+	PM_DEVICE_DT_INST_DEFINE(inst, auxdisplay_jhd1313_pm_action);			\
```
leftover

gmarull commented on dts/bindings/auxdisplay/hit,hd44780.yaml:
```diff
@@ -0,0 +1,86 @@
+
+  enable-line-fall-delay:
+    type: int
+    default: 100
+    description: |
+      Delay time (in us) to wait after enable line falls before sending
+      another command.
+
+  clear-command-delay:
+    type: int
+    default: 5000
+    description: |
+      Delay time (in us) to wait after issuing a clear command before sending
+      another command.
```
to all: defaults need to be justified, per bindings policy

gmarull replied:
also, dt props that have units should contain proper suffix, e.g. `-us`

gmarull reviewed (changes requested):
overall lgtm, spotted a few issues

gmarull commented on drivers/auxdisplay/auxdisplay_jhd1313.c:
```diff
@@ -0,0 +1,369 @@
+};
+
+#define AUXDISPLAY_JHD1313_DEVICE(inst)							\
+	static struct auxdisplay_jhd1313_config auxdisplay_jhd1313_config_##inst = {	\
```
`const`

gmarull commented on drivers/auxdisplay/auxdisplay_hd44780.c:
```diff
@@ -0,0 +1,595 @@
+		.capabilities = {								\
+			.columns = DT_INST_PROP(inst, columns),					\
+			.rows = DT_INST_PROP(inst, rows),					\
+			.mode = DT_ENUM_IDX(DT_DRV_INST(inst), mode),				\
```
DT_INST…

gmarull commented on drivers/auxdisplay/auxdisplay_hd44780.c:
```diff
@@ -0,0 +1,595 @@
+		.clear_delay = DT_INST_PROP(inst, clear_command_delay_us),			\
+		.boot_delay = DT_INST_PROP(inst, boot_delay_ms),				\
+	};											\
+	PM_DEVICE_DT_INST_DEFINE(inst, auxdisplay_hd44780_pm_action);				\
```
leftover

thedjnK replied:
Removed

gmarull commented on samples/drivers/auxdisplay/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+This sample shows how to use the auxiliary display drivers by outputting a
+sample "Hello World" text to one. If power management is enabled when
+building (:kconfig:option:`CONFIG_PM_DEVICE`), it also demonstrates
```
outdated?

thedjnK replied:
Fixed

gmarull commented on samples/drivers/auxdisplay/src/main.c:
```diff
@@ -0,0 +1,78 @@
+
+	k_sleep(K_SECONDS(3));
+
+	pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
```
needs update?

thedjnK replied:
Removed
